### PR TITLE
os/bluestore: AVL-tree & extent - based space allocator 

### PR DESCRIPTION
**Pull request by xiexingguo**
Steal from ZFS :-)
Local (very rough) tests show that the performance is comparable with Stupid(at least), need further verification, though.
Benefits: better memory monitoring, perhaps? Also simpler code logic if you don't have to dig into the details of AVL-implementation...

### CODE REVIEW

**Comment by mattbenjamin** [comment]
@xiexingguo boost::intrusive contains a flexible avl-tree implementation with tree as well as set-oriented interfaces, that's already used in Ceph.  It's not a critique of the submitted work at all, but I think I'd prefer to see us re-use this generic sort of code whenever possible.

Matt

**Comment by mattbenjamin** [comment]
@xiexingguo oh, sorry, you're adding the Sun Microsystems AVL implementation.  That -is- high quality and I've re-used it in projects myself.  There are then 2 points that apply:
1. is there a compelling reason not to use boost?
2. CDDL--I don't know all the ins and out of that;  there has been a claim that (L)GPL and CDDL licensed codes are not compatible;  e.g., https://lwn.net/Articles/688683/;  I don't know the current accepted opinion, to be honest.

**Comment by xiexingguo** [comment]
@mattbenjamin Thanks for your comments.

Actually I have no preference of Sun::AVL than Boost::AVL. I used to be very familiar with these (ZFS) codes, so I just pick the quickest way by copying it here...

If this (AVL allocator) turns out to be a better option or at least an alternative option, I can definitely switch to Boost::AVL.

**Comment by tchaikov** [comment]
i concur with @mattbenjamin , we cannot include CDDL code in our tree in this way. as to my best knowledge, CDDL and LGPL 2.1 is not compatible. the source code of the derivative work (ceph binaries) of LGPL2.1 should be available under the license of LGPL2.1. but the avl implementation is licensed under CDDL.

**Comment by liewegas** [comment]
Can you summarize what the allocator does?  It doesn't seem like the use of an avl tree instead of the rbtree in stupid is the imporant part..

**Comment by xiexingguo** [comment]
> It doesn't seem like the use of an avl tree instead of the rbtree in stupid is the imporant part..

That's true:-)
(1) We manage the space in extent, each extent is of form [offset, offset+length]/[start, end);
(2) We use two avl trees to track all free(unallocated) extents, namely `range-tree` and `range-size-tree`.
    ` range-tree` arranges extents strictly by offset, while the `range-size-tree` arranges extents strictly by the size of each extent. All extents are linked and updated simutaneously in both trees so we won't double the memory consumpation.
(3) Each time we allocate some space, the corresponding range will be removed from a specific matched extent, and that extent will split if necessary.
(4) Each time we free some space, the corresponding range will be added back to both trees, and a merge operation will be automatically performed if necessary.
(5) We always try to search and allocate space from the `range_tree` (in term of **First-Fit**, which is very fast). But if the free space is low or the whole space mag gets too fragmented, we then switch to the `range_size_tree` to use the **Best-Fit** policy, which by definition, always try to find the extent that has the closest size of the requested amount of space.

**Comment by liewegas** [comment]
Sounds good!  We should switch over to the boost instrusive AVL code though :)

File: src/os/bluestore/AvlAllocator.cc [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+      // Allocation failed.
+      break;
+    }
+    block_list.add_extents(offset, length);
+    allocated += length;
+  }
+
+  if (allocated) {
+    std::lock_guard<std::mutex> l(lock);
+    num_reserved -= allocated;
```
Comment (ifed01) [inline src/os/bluestore/AvlAllocator.cc]: let's do that inside _allocate to avoid additional lock acquisition.

File: src/os/bluestore/AvlAllocator.cc [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+ */
+uint64_t AvlAllocator::_block_picker(avl_tree_t *t,
+  uint64_t *cursor, uint64_t size, uint64_t align)
+{
+  range_seg_t *rs = _find_block(t, *cursor, size);
+  while (rs != NULL) {
+    uint64_t offset = P2ROUNDUP(rs->rs_start, align);
+
+    if (offset + size <= rs->rs_end) {
+      *cursor = offset + size;
```
Comment (ifed01) [inline src/os/bluestore/AvlAllocator.cc]: General concern about such cursor behavior (already highlighted for stupid alocator at #18494) is that it permanently increases that causes excessive fragmentation. Here is a scenario I observed for Stupid Allocator and that's probably is applicable here as well:
There is 40Gb block device (coupled with 16 Gb DB and 10 Gb WAL devices but that's doesn't matter). It's sequentially pre-condifitioned with 32 Gb of data. Hence we have 8 Gb free at block device and allocator has a single continuous 8Gb free extent at this point. Then we start doing random 4K overwrites that cause some 4K extents to return to allocator while new ones are allocated from that large 8Gb extent. Please note that returned small extents aren't reused until we drain all the 8Gb extent since cursor permanently goes up. After overwriting 8Gb of data we have large continuous extent fully exhausted and free 8Gb spread randomly over released 4K extents. There are good chances that existing continuous extents are pretty short (<1Mb). One day bluefs rebalance procedure comes into action and requests that long continuous extent(s) of 1Mb... Allocator fails to do that and we have an assert. That's a real use case I faced recently...
May be we should do a cursor rewind on extent release?

File: src/os/bluestore/AvlAllocator.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  static constexpr unsigned MAX_LBAS = 64;
+  uint64_t lbas[MAX_LBAS] = {0};
+
+  /*
+   * Minimum size which forces the dynamic allocator to change
+   * it's allocation strategy.  Once the allocator cannot satisfy
+   * an allocation of this size then it switches to using more
+   * aggressive strategy (i.e search by size rather than offset).
+   */
+  uint64_t range_size_alloc_threshold = 1ULL << 17; ///< 128K
```
Comment (ifed01) [inline src/os/bluestore/AvlAllocator.h]: Suggest to make externally configurable

File: src/os/bluestore/AvlAllocator.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  /*
+   * The minimum free space, in percent, which must be available
+   * in allocator to continue allocations in a first-fit fashion.
+   * Once the allocator's free space drops below this level we dynamically
+   * switch to using best-fit allocations.
+   */
+  int range_size_alloc_free_pct = 4;
```
Comment (ifed01) [inline src/os/bluestore/AvlAllocator.h]: And this one too.

**Comment by liewegas** [comment]
@xiexingguo @ifed01 Should we merge this as is?  Abandon it?  If we let it sit here in this PR it will just rot.

**Comment by xiexingguo** [comment]
> @xiexingguo @ifed01 Should we merge this as is? Abandon it? If we let it sit here in this PR it will just rot.

@liewegas I'd suggest get this one in as it is.  It's a good alternative and we can do further testing and improvements once it gets in.

**Comment by ifed01** [comment]
@liewegas - there were some licensing concerns above.. I know almost nothing about that but perhaps it's better to replace with Boost implementation then.

**Comment by liewegas** [comment]
Oh yeah, I forgot about the CDDL issue.  That needs to be fixed before we can merge this.

**Comment by tchaikov** [comment]
@xiexingguo is this change still relevant ?

**Comment by tchaikov** [comment]
adapted to boost::intrusive::avltree, see #30897
